[Doc] Document Kernel Templates (#1328)
oneapi-src/oneDPL: documentation/library_guide/kernel_templates/esimd_main.rst

ESIMD-Based Kernel Templates
############################

The ESIMD kernel templates are based on `Explicit SIMD SYCL extension <https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-0/explicit-simd-sycl-extension.html>`_ of Intel® oneAPI DPC++/C++ Compiler.
This technology only supports Intel GPU devices.

> The following algorithms are implemented:

Reviewer: Add (probably prepend) something about the ``esimd`` namespace, e.g. "These templates are available in the ``oneapi::dpl::experimental::kt::esimd`` namespace."

Author: Done

> ESIMD-based Kernel Templates

Reviewer 2: Please capitalize based: ESIMD-Based Kernel Templates

Author: Done

############################

> ESIMD Kernel Templates are based on `Explicit SIMD SYCL (ESIMD) <https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-0/explicit-simd-sycl-extension.html>`_.

Reviewer 2: My suggestions (note that length is spelled incorrectly in vector_lenght):

The ESIMD kernel templates are based on `Explicit SIMD SYCL (ESIMD) <https://www.intel.com/content/www/us/en/docs/dpcpp-cpp-compiler/developer-guide-reference/2024-0/explicit-simd-sycl-extension.html>`_. This technology only supports Intel GPU devices.

NOTE:
The ``kernel_param::data_per_workitem`` parameter has a special meaning in ESIMD-based kernel templates. Usually, each work item processes ``data_per_workitem`` sequentially. However, work items in ESIMD-based kernel templates perform vectorization, so the sequential work is ``data_per_workitem / vector_lenght``, where ``vector_lenght`` is an implementation-defined vectorization factor.

Author: I have fixed the length spelling. It seems the rest is the same besides "work-item"/"work item" which we agreed to leave as is.

Reviewer 2: LGTM

   :hidden:

   esimd/radix_sort

-------------------
System Requirements
-------------------

- Hardware: Intel® Data Center GPU Max Series.
- Compiler: Intel® oneAPI DPC++/C++ Compiler 2023.2 and newer.
- Operating Systems:

  - Red Hat Enterprise Linux* 9.2,
  - SUSE Linux Enterprise Server* 15 SP5,
  - Ubuntu* 22.04.

>   - Other distributions and their versions listed in `<https://dgpu-docs.intel.com/driver/installation.html>`_ should be supported accordingly, however they have not been tested.

Reviewer: ```suggestion
  - Other distributions and their versions listed in `<https://dgpu-docs.intel.com/driver/installation.html>`_ (testing may be limited or not performed).
```
I am not sure though if something like this is appropriate at all in our documentatiion.

Author: Let's remove that line entirely. It will have the same what oneAPI system requirements have: https://www.intel.com/content/www/us/en/developer/articles/system-requirements/intel-oneapi-base-toolkit-system-requirements.html#:~:text=Intel%C2%AE%20Data%20Center%20GPU%20Max%20Series%20(PVC,Ubuntu%2022.04
